### PR TITLE
fix(test): rebuild the MSL AD-vs-FD comparator in float64 and gate its resolving power (closes #527)

### LINKED ISSUE #527: MSL AD-vs-FD gate: the f32 single-h FD comparator cannot resolve the 0.10 bound (not an AD defect)
The MSL AD gradient no longer matches finite differences: **`rel_err = 0.8519`** against a 0.10 gate. Measured on merged main (`f95240f`) by the maintained GPU harness.

This is the re-measurement #515 called for after PR #516 replaced the S-matrix assembly. The pre-#516 record was `rel_err = 0.00011`.

## The run

VESSL **369367250602**, `scripts/vessl_gpu_suite.yaml`, `remilab-c0` / gpu-rtx4090, main@`f95240f`, 2026-07-31. Log backed up to `/tmp/gpu_369367250602.log`; full pytest log in the mounted artifact `.omx/gpu-suite/20260731T053033Z-f95240f/gpu_suite.log`.

```
1 failed, 193 passed, 70 skipped, 2912 deselected, 2 xfailed  in 4061.29s (1:07:41)
FAILED tests/test_msl_ad_fd_converged.py::test_msl_ad_fd_converged_tight
```

**One failure in the whole GPU suite**, and it is the gradient gate.

## The numbers

```
[MSL-FD-TIGHT] n_steps=26226, checkpoint_segments=141 (~sqrt=161.9)
[MSL-FD-TIGHT] forward |S| range: [0.0080, 1.0000]
[MSL-FD-TIGHT] loss = 1.600595e+01
[MSL-FD-TIGHT] g_ad = -4.236159e-03  (AD wall-time: 68.5s)
[MSL-FD-TIGHT] g_fd = -2.861023e-02  (FD wall-time: 34.9s, h=0.001)
[MSL-FD-TIGHT] rel_err = 0.8519  (threshold: 0.10)
[MSL-FD-TIGHT] sign agreement: g_ad=-4.2362e-03 g_fd=-2.8610e-02
[MSL-FD-TIGHT] num_periods=20, n_freqs=8
```

| | AD | FD | rel_err |
|---|---|---|---|
| pre-#516 (f64 referee, #486) | −2.11425e-01 | −2.11449e-01 | **0.00011** |
| post-#516 (this run) | **−4.2362e-03** | **−2.8610e-02** | **0.8519** |

Signs still agree, and the forward pass is healthy (`|S|` in [0.008, 1.000], well inside the guard's [0, 1.2]).

**The shape of the change is the informative part.** Both gradients dropped, but not together: FD fell ~7x (0.211 → 0.0286) while AD fell ~50x (0.211 → 0.00424). A changed extractor legitimately changes the loss and therefore both gradients — it does **not** explain them ceasing to agree with each other. The disagreement is the finding.

## What is NOT the explanation

- **Not a broken forward pass** — `|S|` range healthy, loss finite, both gradients nonzero with matching sign.
- **Not the smoke-test vacuity of #515** — that was a synthetic-fixture artifact; this is the real end-to-end gate on a physical fixture.
- **Not gate drift** — the threshold is untouched at 0.10, and the test's own docstring says: *"If rel_err stays >10% this test will fail (deliberately — do NOT loosen the gate to force a pass; report as a gradient accuracy finding instead)."* Following that.

## The lead I would start from

**This fixture is dx = 80 µm, `h_sub/dx` = 3.175 — the bisecting mesh class of #525.** Its preflight, verbatim from the run:

```
[PREFLIGHT] MSL port 'msl_0': h_sub/dx = 3.175 (fractional part 0.175) lands in the
  [0.10, 0.40] mixed-cell danger zone. The substrate-air interface bisects the same
  Yee cell that holds the trace ... To snap onto a safe alignment, set dx = 63.5µm
  (= h_sub/4) or 84.7µm (= h_sub/3).
[PREFLIGHT] MSL port 'msl_0': only 3 substrate cell(s) in z ... Z0 staircase error >5%
```

On that same mesh class, #525 measured the modal V·I pair recovering only **0.54–0.69** of the Poynting flux (vs 1.006–1.009 on an aligned mesh). Two independent failures on the same mesh class is the strongest single clue available, and #525's probe (span sweep × alignment class) is running now.

That said — **two symptoms on one mesh is a correlation, not a mechanism.** A V·I pair that mis-measures power by a fixed factor would move AD and FD *together*, not apart, so #525 cannot be the whole story here. Something in the AD path specifically is losing sensitivity.

## Candidate mechanisms, none tested

1. The multi-drive solve puts `jnp.linalg.solve` on the tape. Unit-level AD-vs-FD through `msl_solve_s_from_waves` was verified at 1.165e-9 (PR #516 review), so the primitive is sound — but its **conditioning** interacts with reverse-mode differently than the old per-column ratio did.
2. Something upstream of the solve became effectively non-differentiable or gradient-attenuating for this parameter (the fixture differentiates an `eps_r_sub` override; #483's lesson was that a launch fixture can silently freeze what the tape sees).
3. The checkpointed segmented scan (`checkpoint_segments=141`) interacting with the new assembly.
4. The FD reference is itself noisy — the test's own docstring records a CPU h-sweep showing 27% non-monotone FD scatter. At `g_fd` = 0.0286 (7x smaller than before) that noise floor is proportionally much larger, so part of the 0.85 could be reference noise. **An h-sweep on the current code is the cheapest way to bound this before chasing AD.**

## Next

Start with (4) — it is cheap and bounds how much of the 0.85 is real. Then (2), because a frozen/attenuated parameter path is the failure mode this repo has hit before (#483) and it predicts exactly this signature: AD far smaller than FD.

Blocks the AD-related claims in `docs/guides/sparameter_support_matrix.md` (the "converged f64 referee 0.000110" bullet, flagged in #520) until resolved. Related: #515, #525, #520.

### PR DESCRIPTION
**Pull request by bk-squared**
## The comparator, not the gradient

`test_msl_ad_fd_converged_tight` compared AD against a central difference whose
two loss evaluations ran in **float32**. Near `loss = 16.00599` one float32 ULP
is `1.9073e-06`, and at the gate's `h = 1e-3` the true difference `2h|g|` is
`8.47e-06` — **4.4 ULP**. The subtraction was returning a count of quantisation
steps, not a derivative.

Measured on the gate's own fixture and h:

| comparator | `g_fd` | `rel_err` | resolving power |
|---|---|---|---|
| float32 (shipped) | −2.861023e-02 | **0.8519** | 4.4 ULP |
| float64 (this PR) | −4.1005380e-03 | **0.0331** | 2.31e+09 ULP |

`g_ad = −4.236159e-03` (gpu-rtx4090) / `−4.242490e-03` (CPU). **The AD is
correct.** This was a comparator defect for the gate's whole lifetime.

The proof is arithmetic, not statistical: a 7-point h-sweep (VESSL
369367250742) returned FD values that were **exact integer multiples of
`ULP/(2h)` at every h** — 14, 1, −30, −7, −31, −52, −76 — and at `h = 5e-4` the
entire measured "gradient" was one single ULP. The sign flip that looked like
physics is which side of a representable value each evaluation landed on. No h
repairs it: 1% resolution needs `h > 0.0225`, where `h²` truncation takes over.

### Why it surfaced now

`sum|S|²` is pinned near the passivity value `2·n_freqs = 16`; the measured loss
is `16.00599`, so **0.037% of the objective is signal** riding on a structural
constant. PR #516 removed extractor error, so that residue — and its derivative
— shrank 50x, from `−2.1143e-01` to `−4.2425e-03`, while the loss magnitude and
therefore the ULP stayed put. The gradient fell off the resolvable end of
float32.

Corroborated by a number recorded **before** #516: the #477 work logged a "27%
NON-monotone FD scatter" on this fixture. The same sweep now reads 2017%, and
`27% × 50 ≈ 1350%` is the same order. Read the other way, the gradient collapse
is evidence that #516 worked — the old gate was substantially differentiating an
extractor artefact.

## Changes

1. **The FD reference runs in float64** under a scoped `enable_x64()`. Never at
   module level: process-global, flips at collection, reds every pytest-split
   shard. `probes.py` already keys accumulator dtype off `jax.config.x64_enabled`,
   so x64 reaches `S` (verified `complex128` / `float64`; scoping verified
   non-leaking across tests).
2. **A comparator-validity assert, placed BEFORE the accuracy gate.** The FD
   span must exceed `_MIN_FD_ULP_SPAN = 1e4` ULP. This is the check whose
   absence let this issue be filed against the AD path — the honest reading of
   `rel_err = 0.85` on a 4.4-ULP reference is *"the instrument is broken"*. Its
   failure message says exactly that and tells the reader not to touch the
   extractor or the tape.
3. **`scripts/msl_ad_fd_f64_referee.py`** — the referee, committed. The f64
   measurement behind the docstring's `rel_err = 0.00011` claim existed only as
   a sentence; no script in the repo reproduced it. Third instance of that class
   in this area (#520, #525).
4. The docstring sentence *"the 0.10 threshold now carries real margin"* is
   replaced by the measurement that falsified it.

**The 0.10 threshold is unchanged. Nothing here loosens a gate.**

## Cost, measured on the lane

A float64 reference sounds expensive; on this lane it is not. f32 AD 59.1s, f64
FD 41.7s and 37.7s per h (two forwards each) — ~19s per f64 forward against ~15s
for f32, whole gate ~140s. The FDTD is memory-bandwidth-bound, so the RTX4090's
1/64 float64 throughput barely shows. On CPU the same forward is ~500s, which is
why this was decided from a run on the lane the gate actually uses (VESSL
369367250775) rather than a local timing.

## Falsifier

`test_comparator_floor_rejects_the_f32_reference_that_caused_527` (fast, no
simulation) pins the new floor from **both** sides: it must reject the exact
float32 configuration that caused this issue (4.4 ULP) and accept the float64
one (2.4e9 ULP), with a margin assert so a future edit that halves the headroom
is visible. If someone later lowers the floor to quiet a red, that test tells
them what they are re-admitting.

## Preflight (quoted, per repo rule)

The fixture emits, unchanged by this PR and identical on both platforms: `'pec'
z-extent 80µm = 1.0 cells — below 1 cell resolution`; `pec_faces={z_lo} creates
an INFINITE PEC boundary AND the geometry contains finite PEC objects`; `all
dielectric(s) ['ro4350b'] are perfectly lossless in an open (CPML) domain`; and
per port `only 3 substrate cell(s) in z … Z0 staircase error >5% expected` and
`h_sub/dx = 3.175 … lands in the [0.10, 0.40] mixed-cell danger zone`. The last
one matters for interpreting the fixture but not for this change: the gate's
existing `s_max <= 1.2` forward guard covers it, and nothing here alters the
geometry.

Closes #527.

`R3: memory=feedback_root_cause_before_gate_change ("compute the discretization/Yee bound … FIRST" — the written root cause is the float32 ULP quantisation, measured before any gate edit) + feedback_gate_can_bind_artifact (an ADDED gate needs a falsifier) | R2-attempts=0 (comparator repair after a closed root cause, not a mechanism retry) | falsifier=test_comparator_floor_rejects_the_f32_reference_that_caused_527 pins the floor from both sides; note the usual domain-size invariance witness does not apply here — the added gate is arithmetic on the comparator's resolving power, not a physics claim`

### CODE REVIEW

**Comment by bk-squared** [comment]
## The gate runs green on the lane it is marked for

This was the gap I flagged against my own PR: `gpu`+`slow` means CI never executes this test, so it shipped without end-to-end evidence. VESSL **369367250794** runs the pytest file itself — not the referee script, which exercises the same physics but none of the test's asserts or their ordering — on `remilab-c0`/`gpu-rtx4090` at `9bbacbd`.

```
[MSL-FD-TIGHT] loss = 1.600595e+01
[MSL-FD-TIGHT] g_ad = -4.236159e-03  (AD wall-time: 55.8s)
[MSL-FD-TIGHT] g_fd = -4.100538e-03  (FD wall-time: 41.8s, h=0.001)
[MSL-FD-TIGHT] FD reference resolving power: 2.31e+09 ULP of the loss (floor 1e+04)
[MSL-FD-TIGHT] rel_err = 0.0331 (threshold: 0.10, enforced below)
[MSL-FD-TIGHT] sign agreement: g_ad=-4.2362e-03 g_fd=-4.1005e-03
[MSL-FD-TIGHT] total wall-time: 117.5s
[MSL-FD-TIGHT] PASS
3 passed in 119.51s (0:01:59)
```

Three things this establishes that nothing before it did:

1. **x64 actually engaged on the GPU.** The dtype assert sits inside the measured path and did not fire. Before that assert existed, a green run proved nothing — the resolving-power number read ~2.4e9 whether the loss was float32 or float64.
2. **The comparator floor passes on real data**, 2.31e+09 against a 1e+04 floor, with the check running *before* the accuracy gate as designed.
3. **The print/enforce mismatch is gone** — `threshold: 0.10, enforced below`, where it used to print 0.05 while enforcing 0.10.

Cost: 117.5s total, against ~89s for the f32 version it replaces.

A CPU run of the same file is in flight as a second platform (`g_ad = -4.242490e-03`, matching GPU to 0.15%); it is supporting evidence, not the authority — the lane run is.

**Preflight, quoted verbatim** (unchanged by this PR, identical on both platforms): `'pec' z-extent 80µm = 1.0 cells — below 1 cell resolution`; `pec_faces={z_lo} creates an INFINITE PEC boundary AND the geometry contains finite PEC objects`; `all dielectric(s) ['ro4350b'] are perfectly lossless in an open (CPML) domain`; and per port `only 3 substrate cell(s) in z (h_sub=254µm, dx=80µm). Yee staircase at dielectric interface is O(dx) — Z0 staircase error >5% expected` and `h_sub/dx = 3.175 (fractional part 0.175) lands in the [0.10, 0.40] mixed-cell danger zone`.

Logs archived: `docs/research_notes/vessl_logs/20260801_369367250794_msl_ad_gate_529.log`, alongside the referee (369367250775), the h-sweep (369367250742), and the local CPU referee.
